RavenDB-7503

talweiss1982 opened the pull request:
* Now allowing api keys with '/' as long as they are starting with Raven/ and properly split their secret.
* Opening an async task listening to a named pipe constructed as raven-control-pipe-<raven server pid>
* Added a method to create a system api key named Raven/System to be used by the admin in harden environment
* Added trust/init methods to RVN tool, still need TLC will open another issue to have better error handling and validation in RVN

Commit:
RavenDB-7503
* Now allowing api keys with '/' as long as they are starting with Raven/ and properly split their secret.
* Opening an async task listening to a named pipe constructed as raven-control-pipe-<raven server pid>
* Added a method to create a system api key named Raven/System to be used by the admin in harden environment
* Added trust/init methods to RVN tool, still need TLC will open another issue to have better error handling and validation in RVN

ppekrol commented on src/Raven.Server/ServerWide/ServerStore.cs:
```diff
@@ -1395,12 +1398,78 @@ public DynamicJsonValue GetLogDetails(TransactionOperationContext context, int m
         }
 
+        private const string SystemApiKey = "Raven/System";
+        public async Task<(string api_key, string public_key)> GetApiKeyAndPublicKey()
```
@talweiss1982 can you use C# casing here? Task<(string ApiKey, string PublicKey)>

ayende replied:
Please use PascalCase here.

ayende commented on src/Raven.Server/RavenServer.cs:
```diff
@@ -119,7 +121,10 @@ public void Initialize()
             sp.Restart();
-
+            if (Pipe != null)
+            {
```
Add comment explaining why this is.
This is the second time I'm seeing this code and going `HuGH?`

ayende commented on src/Raven.Server/RavenServer.cs:
```diff
@@ -184,6 +189,95 @@ public void Initialize()
             }
         }
 
+        private async Task ListenToPipe()
```
Missing error handling here.

ayende commented on src/Raven.Server/RavenServer.cs:
```diff
@@ -184,6 +189,95 @@ public void Initialize()
+                    }
+                }
+                //From what i read we should not re-use a pipe.
+                Pipe.Dispose();
```
Failure to open / dispose would crash the process

ayende commented on src/Raven.Client/Http/OAuth/ApiKeyAuthenticator.cs:
```diff
@@ -176,7 +176,13 @@ private string[] ExtractApiKeyAndSecret(string apiKey)
             {
-                apiKeyParts[1] = string.Join("/", apiKeyParts.Skip(1));
+                var secretIndex = 1;
+                if (apiKeyParts[0] == "Raven" && apiKeyParts.Length > 3)
```
Maybe better of to use "Raven:Cluster/Secret" ?

ayende replied:
And then reserve just `Raven:`

ppekrol replied:
I think that 'Raven/' is already reserved: https://github.com/ravendb/ravendb/blob/v4.0/src/Raven.Server/Web/Authentication/AdminApiKeysHandler.cs#L26

Commit:
RavenDB-7503
reverting the acceptance of Raven/ will use Raven:System/<Secret>
added try catch on resetting the pipe
added an endpoint to restart the pipe and start listening to it (for the case the pipe reset throws)

ayende commented on src/Raven.Server/RavenServer.cs:
```diff
@@ -189,82 +190,96 @@ public void Initialize()
+        internal async Task ListenToPipe()
         {
-            while (true)
+            lock (this)
```
What, why? This is going to be locking effectively forever.

ayende commented on src/Raven.Server/Web/Authentication/AdminApiKeysHandler.cs:
```diff
@@ -68,6 +68,13 @@ public class AdminApiKeysHandler : AdminRequestHandler
             }
         }
 
+        [RavenAction("/admin/server/reset-pipe", "GET", "/admin/server/reset-pipe")]
```
Why do we need this?

ayende commented:
We are going with a different route here, see: http://issues.hibernatingrhinos.com/issue/RavenDB-7696
